tiny_dnn/core/kernels/recurrent_cell_op_internal.h

```diff
@@ -0,0 +1,152 @@
+                                       tensor_t &curr_output_delta,
+                                       tensor_t &curr_state_delta,
+                                       tensor_t &prev_output_delta,
+                                       tensor_t &prev_state_delta,
+                                       const tensor_t &out_h,
+                                       const recurrent_cell_params &params,
+                                       const bool layer_parallelize) {
+  for (serial_size_t sample = 0; sample < prev_out.size(); sample++) {
+    // from output to h
+    for (serial_size_t o = 0; o < params.out_size_; o++) {
+      // propagate delta from output to h.
+      curr_state_delta[sample][o] +=
+        vectorize::dot(&curr_output_delta[sample][0], &V[o * params.out_size_],
+                       params.out_size_);
+    }
+
+    // h'(t)
+    params.activation_->backward_activation(prev_h[sample], out_h[sample],
+                                            curr_state_delta[sample],
+                                            curr_state_delta[sample]);
+
+    // \delta h(t) -W-> h(t-1)
+    for (serial_size_t o = 0; o < params.out_size_; o++) {
+      prev_state_delta[sample][o] +=
+        vectorize::dot(&curr_state_delta[sample][0], &W[o * params.out_size_],
+                       params.out_size_);
+    }
+
+    // \delta h(t) -U-> \delta x(t)
+    for (serial_size_t i = 0; i < params.in_size_; i++) {
+      prev_output_delta[sample][i] +=
+        vectorize::dot(&curr_state_delta[sample][0], &U[i * params.out_size_],
+                       params.out_size_);
+    }
+
+    for_(layer_parallelize, 0, size_t(params.out_size_),
+         [&](const blocked_range &r) {
+           // accumulate weight-step using delta
+           // dW[c * out_size + i] += current_delta[i] * prev_out[o]
+           for (serial_size_t o = 0; o < params.out_size_; o++) {
+             vectorize::muladd(&curr_output_delta[sample][r.begin()],
+                               out_h[sample][o], r.end() - r.begin(),
+                               &dV[sample][o * params.out_size_ + r.begin()]);
+           }
+
+           if (params.has_bias_) {
+             // vec_t& dc;
+             for (serial_size_t o = r.begin(); o < r.end(); o++) {
+               dc[sample][o] += curr_output_delta[sample][o];
+             }
+           }
+
+           for (serial_size_t o = 0; o < params.out_size_; o++) {
+             vectorize::muladd(&curr_state_delta[sample][r.begin()],
+                               prev_h[sample][o], r.end() - r.begin(),
+                               &dW[sample][o * params.out_size_ + r.begin()]);
+           }
+
+           for (serial_size_t i = 0; i < params.in_size_; i++) {
+             vectorize::muladd(&curr_state_delta[sample][r.begin()],
+                               prev_out[sample][i], r.end() - r.begin(),
+                               &dU[sample][i * params.out_size_ + r.begin()]);
+           }
+
```

Comment:
You shouldn't repeatedly use `r.end() - r.begin()` inside loops. Use a local variable to hold the value.